test/performance/perf-upsert/src/main.rs

```diff
@@ -0,0 +1,160 @@
+// Copyright Materialize, Inc. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Performance test for Materialize Upsert sources
+
+#![deny(missing_debug_implementations, missing_docs)]
+
+use std::process;
+use std::sync::Arc;
+use std::time::Duration;
+
+use anyhow::Result;
+use rand::Rng;
+use structopt::StructOpt;
+
+use test_util::kafka::kafka_client;
+
+use crate::config::{Args, KafkaConfig, MzConfig};
+use crate::mz_client::MzClient;
+
+mod config;
+mod mz_client;
+
+#[tokio::main]
+async fn main() {
+    if let Err(e) = run().await {
+        println!("ERROR: {}", e);
+        let mut err = e.source();
+        while let Some(e) = err {
+            println!("    caused by: {}", e);
+            err = e.source();
+        }
+        process::exit(1);
+    }
+}
+
+async fn run() -> Result<()> {
+    let config = Args::from_args();
+    env_logger::init();
+
+    let k_config = config.kafka_config();
+    let mz_config = config.mz_config();
+
+    log::info!(
+        "starting up message_count={} mzd={}:{} kafka={} preserve_source={}",
+        config.message_count,
+        config.materialized_host,
+        config.materialized_port,
+        config.kafka_url(),
+        config.preserve_source,
+    );
+
+    let k = tokio::spawn(async move { create_kafka_messages(k_config).await });
+    let mz = tokio::spawn(async move { create_materialized_source(mz_config).await });
+    let (k_res, mz_res) = futures::join!(k, mz);
+    k_res??;
+    mz_res??;
+
+    Ok(())
+}
+
+async fn create_kafka_messages(config: KafkaConfig) -> Result<()> {
+    use rand::SeedableRng;
+    let mut rng = rand::rngs::StdRng::from_seed(rand::random());
+    let mut total_messages_sent = 0;
+
+    let val_a: Vec<u8> = "a".repeat(500).into_bytes();
+
+    let k_client = Arc::new(kafka_client::KafkaClient::new(
+        &config.url,
+        &config.group_id,
+        &config.topic,
+        &[],
+    )?);
+
+    if let Some(create_topic) = &config.create_topic {
+        let configs = vec![
```

Comment:
note that we could easily have this be a string of the form "key=value,key2=value2,..." thats sent in from the command line and parsed